### PR TITLE
Chapter 11 - Add Authentication 

### PR DESCRIPTION
**Pull request by ViolaRafael**
Added option to register and login using email saved in database,
Added option to logout,
Changed Favorite Items and Orders to be per user,
Added auto login using shared_preferences,
Added auto log out using a timer from database.

### CODE REVIEW

**Review by PedroRaposoPMA — DISMISSED** [review]

File: lib/components/auth_form.dart [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  bool _isLogin() => _authMode == AuthMode.login;
+  bool _isSignUp() => _authMode == AuthMode.signup;
+
+  void _switchAuthMode() {
+    setState(() {
+      if (_isLogin()) {
+        _authMode = AuthMode.signup;
+      } else {
+        _authMode = AuthMode.login;
+      }
```
Comment (samuelpedropma) [inline lib/components/auth_form.dart]: You change this to something like this:

```dart
_authMode = _isLogin() ? AuthMode.signup : AuthMode.login;
```

File: lib/data/store.dart [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  static Future<bool> saveMap(String key, Map<String, dynamic> value) async {
+    return saveString(key, jsonEncode(value));
+  }
+
+  static Future<String> getString(String key, [String defaultValue = '']) async {
+    final prefs = await SharedPreferences.getInstance();
+    return prefs.getString(key) ?? defaultValue;
+  }
+  static Future<Map<String, dynamic>> getMap(String key) async {
```
Comment (samuelpedropma) [inline lib/data/store.dart:23]: Format the document and don't forget the documentation

File: lib/pages/auth_or_home_page.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import 'package:shop/pages/products_overview_page.dart';
+import '../models/auth.dart';
+
+class AuthOrHomePage extends StatelessWidget {
+  const AuthOrHomePage({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    Auth auth = Provider.of(context);
+    //return auth.isAuth ? ProductsOverviewPage() : AuthPage();
```
Comment (samuelpedropma) [inline lib/pages/auth_or_home_page.dart]: Remove since it's not necessary

**Review by samuelpedropma — CHANGES_REQUESTED** [review]

File: lib/components/auth_form.dart (hunk 1 above)
Comment (ViolaRafael) [reply lib/components/auth_form.dart]: changed file and it still works as intended

File: lib/data/store.dart (hunk 2 above)
Comment (ViolaRafael) [reply lib/data/store.dart:23]: added requested documentation

File: lib/pages/auth_or_home_page.dart (hunk 3 above)
Comment (ViolaRafael) [reply lib/pages/auth_or_home_page.dart]: removed

**Review by samuelpedropma — APPROVED** [review]

**Pull request merged**